src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/ldap/LdapConnection.cs

```diff
@@ -634,30 +634,39 @@ private int SendRequestHelper(DirectoryRequest request, ref int messageID)
                     }
 
                     // Process the attribute.
-                    string stringValue = null;
-                    if (assertion[0] is byte[] byteArray)
+                    byte[] byteArray;
+                    if (assertion[0] is string str)
                     {
-                        if (byteArray != null && byteArray.Length != 0)
-                        {
-                            berValuePtr = new BerVal
-                            {
-                                bv_len = byteArray.Length,
-                                bv_val = Marshal.AllocHGlobal(byteArray.Length)
-                            };
-                            Marshal.Copy(byteArray, 0, berValuePtr.bv_val, byteArray.Length);
-                        }
+                        var encoder = new UTF8Encoding();
+                        byteArray = encoder.GetBytes(str);
+                    }
+                    else if (assertion[0] is Uri uri)
+                    {
+                        var encoder = new UTF8Encoding();
+                        byteArray = encoder.GetBytes(uri.ToString());
+                    }
+                    else if (assertion[0] is byte[] bytes)
+                    {
+                        byteArray = bytes;
                     }
                     else
                     {
-                        stringValue = assertion[0].ToString();
+                        throw new ArgumentException(SR.ValidValueType);
                     }
 
+                    berValuePtr = new BerVal
+                    {
+                        bv_len = byteArray.Length,
+                        bv_val = Marshal.AllocHGlobal(byteArray.Length)
+                    };
+                    Marshal.Copy(byteArray, 0, berValuePtr.bv_val, byteArray.Length);
```

Comment:
I wonder if `byte[] byteArray` is needed or if this can be stackallocated (and the copied)?

Reply:
This code follows how BuildAttributes method does the same conversions. I don't think improving this is within scope of this fix.

Reply:
> improving this is within scope of this fix.

You are right, that's going beyond the fix.

The code here seems pretty old, and was written before span came around. So there could be a general overhaul -- but I don't know the plans here and if it is worth, as such an upgrade comes with cost and this code seems to be battle-tested, which a re-write has to prove first.

Reply:
I don't think we should stackalloc this particular array since it is based on user input, so we want to make sure that we don't call stackalloc from arbitriary input size.

Reply:
Using something like Marshal.StringToCoTaskMemUTF8 could save one allocation. But as I said I don't think this should be within scope of this fix.